docs/prover/approx/grounding.md

```diff
@@ -0,0 +1,274 @@
+
+```cvl
+require    exists address x . forall uint y . e(x,y);
+assert   !(exists address x . forall uint y . e(x,y));
+require  !(forall address x . exists uint y . e(x,y));
+assert !(!(forall address x . exists uint y . e(x,y)));
+```
+
+but these are disallowed:
+
+```cvl
+require  forall address x . exists uint y . e(x,y);
+assert !(forall address x . exists uint y . e(x,y));
+```
+
+One common way to work around these limitations is by replacing `exists`
+expressions with concrete expressions that produce the values that should exist.
+
+For example, suppose your contract function always returned twice its input:
+
+```solidity
+function f(uint x) external returns(uint) { return 2 * x; }
+```
+
+You might like to prove that if `x` is positive, then it's possible for `f` to
+output something smaller than `f(x)`:
+
+```cvl
+assert forall uint x . (x > 0) => (exists uint y . f(y) < f(x));
+```
+
+This is clearly true; for example `f(x - 1)` is always smaller than `f(x)`, as
+is `f(0)`.  However, to work around the nested quantifier restriction, we have
+to help the Prover find the correct value for `y`.  We could replace this
+statement with either of the following two:
+
+```cvl
+assert forall uint x . (x > 0) => f(0) < f(x);
+assert forall uint x . (x > 0) => f(x-1) < f(x);
+```
+
+(grounding-recursion)=
+### Recursion
+
+Quantified statements that relate a function with itself on two different
+inputs may give incorrect counterexamples.  For example, the following `forall`
+statement refers to `f` twice:
+
+```
+rule recursiveQuantifier {
+    require forall uint x . f(x) > f(x-1);
+    assert f(8) > f(6);
+}
+```
+
+Although we can see that the assertion must be true, we would need to combine
+the statements `f(8) > f(7)` and `f(7) > f(6)` to prove it, and the grounding
+mechanism is unable to do this.
+
+In these cases, you may see a counterexample that doesn't satisfy the `require`
+statement.
+
+```{todo}
+Do we have general suggestions for rewriting these?
+```
+
+(grounding-arguments)=
+### Variables must be arguments
+
+In order for grounding to work, every variable appearing in a quantified
+statement must be used at least once as an argument to a contract function.
+For example, neither of the following examples are allowed:
+
+```cvl
+require forall mathint x . x * 2 != y;
+require forall uint x . forall mathint y . f(x) != 2 * y;
+```
+
+In the first example, `x` is not used as an argument to a function, while in the
+second case, `y` is not.
+
+
+Although you are allowed to call functions on complicated expressions that use
+quantified variables, doing so may produce spurious counterexamples.  For
+example, the following is allowed, but is likely to produce spurious
+counterexamples (because `x` itself is not an argument to a function):
+
+```{todo}
+I'm sort of surprised that we allow this.  It would be simpler if we just required
+that `x` must be used as a simple argument to some function, right?
+```
```

Comment:
I'm not sure we don't. But we probably should, but there might be caveats, like cases where we don't really care about the `x`. We actually generate some quantified expressions automatically, and they may not always be tight, i.e., may contain such cases. 

But it's worth a check.